File: class26/java/TCC2601-Main.java
// TCC2601-Main.java | Aula 26
// Driver Java para a gramática ANTLR TCC2601-Expr.g4.
// Avalia expressões usando Visitor.
// Build: antlr4 -Dlanguage=Java -visitor TCC2601-Expr.g4
//        javac TCC2601Expr*.java TCC2601Main.java
//        java TCC2601Main "3 + 4 * 2"
import org.antlr.v4.runtime.*;
import org.antlr.v4.runtime.tree.*;

public class TCC2601Main extends TCC2601ExprBaseVisitor<Double> {

    @Override
    public Double visitProg(TCC2601ExprParser.ProgContext ctx) {
        return visit(ctx.expr());
    }

    @Override
    public Double visitAdd(TCC2601ExprParser.AddContext ctx) {
        return visit(ctx.expr(0)) + visit(ctx.expr(1));
    }

    @Override
    public Double visitMul(TCC2601ExprParser.MulContext ctx) {
        return visit(ctx.expr(0)) * visit(ctx.expr(1));
    }

    @Override
    public Double visitParens(TCC2601ExprParser.ParensContext ctx) {
        return visit(ctx.expr());
    }

    @Override
    public Double visitNumber(TCC2601ExprParser.NumberContext ctx) {
        return Double.parseDouble(ctx.NUM().getText());
    }

    public static void main(String[] args) {
        String input = args.length > 0 ? args[0] : "3 + 4 * 2";
        System.out.println("=== ANTLR Java Driver ===");
        System.out.println("Expressão: " + input);
        CharStream cs = CharStreams.fromString(input);
        TCC2601ExprLexer lexer = new TCC2601ExprLexer(cs);
        CommonTokenStream tokens = new CommonTokenStream(lexer);
        TCC2601ExprParser parser = new TCC2601ExprParser(tokens);
        ParseTree tree = parser.prog();
        TCC2601Main visitor = new TCC2601Main();
        Double result = visitor.visit(tree);
        System.out.println("Resultado: " + result);
        System.out.println("Árvore: " + tree.toStringTree(parser));
    }
}

File: class26/python/TCC2601-main.py
#!/usr/bin/env python3
"""TCC2601-main.py | Aula 26
Driver Python para a gramática ANTLR TCC2601-Expr.g4.
Avalia expressões usando Visitor — mesmo comportamento do driver Java.
Build: antlr4 -Dlanguage=Python3 -visitor TCC2601-Expr.g4
       python3 TCC2601-main.py "3 + 4 * 2"
"""
import sys
from antlr4 import *
# Descomente após gerar o parser com antlr4:
# from TCC2601ExprLexer import TCC2601ExprLexer
# from TCC2601ExprParser import TCC2601ExprParser
# from TCC2601ExprVisitor import TCC2601ExprVisitor


class EvalVisitor:  # (TCC2601ExprVisitor)
    """Visitor que avalia expressões — espelho do driver Java."""

    def visitProg(self, ctx):
        return self.visit(ctx.expr())

    def visitAdd(self, ctx):
        return self.visit(ctx.expr(0)) + self.visit(ctx.expr(1))

    def visitMul(self, ctx):
        return self.visit(ctx.expr(0)) * self.visit(ctx.expr(1))

    def visitParens(self, ctx):
        return self.visit(ctx.expr())

    def visitNumber(self, ctx):
        return float(ctx.NUM().getText())

    def visit(self, ctx):
        """Despacho manual (em produção, herdar de TCC2601ExprVisitor)."""
        method = getattr(self, f"visit{type(ctx).__name__[:-7]}", None)
        if method:
            return method(ctx)
        return None


def main():
    entrada = sys.argv[1] if len(sys.argv) > 1 else "3 + 4 * 2"
    print("=== ANTLR Python Driver ===")
    print(f"Expressão: {entrada}")

    # Descomente após gerar o parser:
    # input_stream = InputStream(entrada)
    # lexer = TCC2601ExprLexer(input_stream)
    # tokens = CommonTokenStream(lexer)
    # parser = TCC2601ExprParser(tokens)
    # tree = parser.prog()
    # visitor = EvalVisitor()
    # result = visitor.visit(tree)
    # print(f"Resultado: {result}")
    # print(f"Árvore: {tree.toStringTree(recog=parser)}")

    # Demonstração sem o parser gerado:
    print("(gere o parser com: antlr4 -Dlanguage=Python3 -visitor TCC2601-Expr.g4)")


if __name__ == "__main__":
    main()
